src/Grpc.AspNetCore.Server/Internal/StreamUtils.cs

```diff
@@ -0,0 +1,140 @@
+
+            var compressionFlag = delimiterBuffer[0];
+            var messageLength = DecodeMessageLength(new ReadOnlySpan<byte>(delimiterBuffer, 1, MessageDelimiterSize));
+
+            if (compressionFlag != 0)
+            {
+                // TODO(jtattermusch): support compressed messages
+                throw new IOException("Compressed messages are not yet supported.");
+            }
+
+            var msgBuffer = new byte[messageLength];
+            if (!await ReadExactlyBytesOrNothing(stream, msgBuffer, 0, msgBuffer.Length))
+            {
+                throw new IOException("Unexpected end of stream.");
+            }
+            return msgBuffer;
+        }
+
+        public static async Task WriteMessageAsync(Stream stream, byte[] buffer, int offset, int count, bool flush = false)
+        {
+            // Write message to the stream with one call
+            // HTTP/2 will place separate calls to Stream.WriteAsync in their own DATA sections in packet
+            // Switching to pipelines will give greater control
+            var completeBuffer = new byte[1 + MessageDelimiterSize + count];
+            completeBuffer[0] = 0; // = non-compressed
+            EncodeMessageLength(count, new Span<byte>(completeBuffer, 1, MessageDelimiterSize));
+            Array.Copy(buffer, offset, completeBuffer, 5, count);
+
+            await stream.WriteAsync(completeBuffer, 0, completeBuffer.Length);
+
+            if (flush)
+            {
+                await stream.FlushAsync();
+            }
+        }
+
+        // reads exactly the number of requested bytes (returns true if successfully read).
+        // returns false if we reach end of stream before successfully reading anything.
+        // throws if stream ends in the middle of reading.
+        private static async Task<bool> ReadExactlyBytesOrNothing(Stream stream, byte[] buffer, int offset, int count)
+        {
+            bool noBytesRead = true;
+            while (count > 0)
+            {
+                int bytesRead = await stream.ReadAsync(buffer, offset, count);
+                if (bytesRead == 0)
+                {
+                    if (noBytesRead)
+                    {
+                        return false;
+                    }
+                    throw new IOException("Unexpected end of stream.");
+                }
+                noBytesRead = false;
+                offset += bytesRead;
+                count -= bytesRead;
+            }
+            return true;
+        }
+
+        private static void EncodeMessageLength(int messageLength, Span<byte> dest)
+        {
+            if (dest.Length < MessageDelimiterSize)
+            {
+                throw new ArgumentException("Buffer too small to encode message length.");
+            }
+
+            ulong unsignedValue = (ulong)messageLength;
+            for (int i = MessageDelimiterSize - 1; i >= 0; i--)
+            {
+                // msg length stored in big endian
```

Comment:
Does this assume the host is little-endian? `System.Net.IPAddress.HostToNetworkOrder` maybe then?

Reply:
As far as I know this is written to work regardless of the endianess of the host.

Writing the message length will be updated to use `BinaryPrimitives.WriteUInt32BigEndian` and `BinaryPrimitives.ReadUInt32BigEndian` in a later PR.

Reply:
https://github.com/grpc/grpc/blob/2b99c000f1d549f2e184968f5b0b34c40d7e3b7b/doc/PROTOCOL-HTTP2.md#requests It's defined as big endian in the document.